### PR TITLE
[BugFix] remove previous worker when modify backend

### PR DESCRIPTION
**Pull request by starrocks-xupeng**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java [hunk 1]
```diff
@@ -175,6 +175,36 @@ public void testAddAndRemoveWorker() throws Exception {
...
+            }
+        };
+
+        long backendId = 5;
+        Deencapsulation.setField(starosAgent, "serviceId", "1");
+        starosAgent.addWorker(backendId, workerHost);
+        Assert.assertEquals(workerId1, starosAgent.getWorkerIdByBackendId(backendId));
+
+        final String workerHost2 = "127.0.0.1:8091";
+        new Expectations() {
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java:197]: add exepctation, expect following call will be made
```
client.removeWorker("1", workerId1)
```

File: fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java [hunk 2]
```diff
@@ -216,12 +216,34 @@ public void addWorker(long backendId, String workerIpPort) {
...
+    private void tryRemovePreviousWorker(long backendId) {
+        long prevWorkerId = getWorkerIdByBackendIdInternal(backendId);
+        if (prevWorkerId < 0) {
+            return;
+        }
+        try {
+            client.removeWorker(serviceId, prevWorkerId);
+        } catch (StarClientException e) {
+            // TODO: fix this corner case later in star mgr
+            LOG.error("Failed to remove worker. error: {}", e.getMessage());
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java:236]: log the worker id as well?

File: fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java [hunk 3]
```diff
@@ -175,6 +175,36 @@ public void testAddAndRemoveWorker() throws Exception {
...
 
+    @Test
+    public void testAddWillRemovePreviousWorker() throws Exception {
+        final String workerHost = "127.0.0.1:8090";
+        final long workerId1 = 10;
+        final long workerId2 = 11;
+        new Expectations() {
+            {
+                client.addWorker("1", workerHost);
+                minTimes = 0;
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java:186]: minTimes = 1;

File: fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java (hunk 1 above)
Comment (starrocks-xupeng) [reply fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java:197]: no need to do that, we can't trace the remove in test, i tested it manually

File: fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java (hunk 3 above)
Comment (starrocks-xupeng) [reply fe/fe-core/src/test/java/com/starrocks/lake/StarOSAgentTest.java]: done

File: fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java (hunk 2 above)
Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java]: done

**Review by kevincai — DISMISSED** [review]

**Review by sduzh — DISMISSED** [review]

**Comment by starrocks-xupeng** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java [hunk 4]
```diff
@@ -216,12 +216,34 @@ public void addWorker(long backendId, String workerIpPort) {
...
+        try {
+            client.removeWorker(serviceId, prevWorkerId);
+        } catch (StarClientException e) {
+            // TODO: fix this corner case later in star mgr
+            LOG.error("Failed to remove worker {} with backend id {}. error: {}", prevWorkerId, backendId, e.getMessage());
+        }
+        workerToBackend.remove(prevWorkerId);
+        for (Map.Entry<String, Long> entry : workerToId.entrySet()) {
+            if (entry.getValue() == prevWorkerId) {
+                workerToId.remove(entry.getKey());
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java:241]: ConcurrentModificationException?

Comment (starrocks-xupeng) [reply fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java:241]: this private method is already protected by outer lock

Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/lake/StarOSAgent.java:241]: I mean remove item when iterate map.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 19 / 22 (86.36%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/StarOSAgent.java|19|22|86.36%|[234, 236, 244]|

**Review by kevincai — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by starrocks-xupeng** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/StarOSAgent.java|16|18|88.89%|[234, 236]|

**Review by sduzh — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**